### PR TITLE
8343064: ClassFormatError: Illegal class name from InnerClassLambdaMetafactory

### PR DESCRIPTION
**Pull request by liach**
After the ClassFile API migration, when serializable lambdas are requested for hidden class callers, illegal class name is generated for the serialization methods, which previously had legal yet unusable class names, as hidden classes cannot be described by a `CONSTANT_Class_info`.

This patch restores a similar older behavior of using legal yet unusable class names.  Previously, the invalid `.` in hidden class names was replaced by a `/` as if a package separator; this patch changes it to a `_` like that in the generated lambda's class name.

The bug report shows some unintended and erroneous usages of `LambdaMetafactory`.  To clarify and to persuade against misuse, I added a paragraph of API notes to `LambdaMetafactory`, describing the impact of this API being designed to support Java language features.  In particular, I used the scenario where people assumed LMf generates weak hidden classes, which happened in this issue report and in #12493, that misuse can lead to resource leaks.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8343064](https://bugs.openjdk.org/browse/JDK-8343064): ClassFormatError: Illegal class name from InnerClassLambdaMetafactory (**Bug** - P3)

### Reviewers
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**)

### Reviewers without OpenJDK IDs
 * @abdelhak-zaaim (no known openjdk.org user name / role) Review applies to [08086c42](https://git.openjdk.org/jdk/pull/21912/files/08086c42f9c75a72032e8ed3aeed5e37b0e08053)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21912/head:pull/21912` \
`$ git checkout pull/21912`

Update a local copy of the PR: \
`$ git checkout pull/21912` \
`$ git pull https://git.openjdk.org/jdk.git pull/21912/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21912`

View PR using the GUI difftool: \
`$ git pr show -t 21912`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21912.diff">https://git.openjdk.org/jdk/pull/21912.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21912#issuecomment-2457902212)
</details>

### CODE REVIEW

**Review by abdelhak-zaaim — APPROVED** [review]

**Comment by AlanBateman** [comment]
Is it possible to add a test for this?

**Comment by liach** [comment]
Test added to ensure that serializable lambdas can be created and its SAM can be called when the caller is a hidden class.  Note that we do not guarantee anything about the serialization functionality itself.  We can add serialization support for hidden classes in the future, but I doubt we do as it is niche and requires larger implementation changes.

**Comment by liach** [comment]
I have updated the documentation.

The API note has been simplified to a simple warning about the intended uses and the resource leak risk.

A new implementation note is added, mentioning the notorious "can't be unloaded" behavior #12493 in the reference implementation explicitly, and describe its implications.

File: src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java [hunk 1]
```diff
@@ -242,6 +229,20 @@
...
+ * support, and such changes may impact unintended uses.  Unintended uses of
+ * these linkage methods may lead to resource leaks, or other unspecified
+ * negative effects.
+ *
+ * @implNote In the reference implementation, the classes implementing the created
+ * function objects are strongly reachable from the defining class loader of the
+ * caller.  This technique reduces heap memory use, but it prevents the
+ * implementation class from unloading when the caller is a {@linkplain
+ * MethodHandles.Lookup.ClassOption#STRONG hidden class} and is unloaded, or when
+ * the returned call site is no longer strongly reachable.
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java:244]: I don't get the last part of the sentence here. When the call site is no longer reachable _that_ prevents the class from being unloaded?

I like the implNote though. Maybe: "This technique reduces heap memory use, but it prevents the implementation class from being unloaded independently of its defining class loader" ?

Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java:244]: Re this sentence: I wish to indicate "caller is unloaded and impl class is unreachable elsewhere (besides in the ClassLoader)".  Maybe just "caller is unloaded" is fine. Thoughts on phrasing?

Comment (JornVernee) [reply src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java:244]: Maybe 'caller's class is unloaded'.

Also, is the link correct? It links to the STRONG option, but the link text is 'hidden class'.

Comment (liach) [reply src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java:244]: The link is intended: I wish the link to describe the unloadable scenario for hidden classes.

File: src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java [hunk 2]
```diff
@@ -229,6 +229,19 @@
...
+ * any time to better suit the Java language features they were designed to
+ * support, and such changes may impact unintended uses.  Unintended uses of
+ * these linkage methods may lead to resource leaks, or other unspecified
+ * negative effects.
+ *
+ * @implNote In the reference implementation, the classes implementing the created
+ * function objects are strongly reachable from the defining class loader of the
+ * caller.  This technique reduces heap memory use, but it prevents the
+ * implementation class from unloading when the caller is a {@linkplain
+ * MethodHandles.Lookup.ClassOption#STRONG hidden class} and is unloaded.
```
Comment (JornVernee) [inline src/java.base/share/classes/java/lang/invoke/LambdaMetafactory.java:243]: Sorry but, the text still doesn't seem right to me. The caller class is unloaded when its class loader is unloaded, which would also be when the implementation class is unloaded (its tied to the caller's class loader). Whether the caller is a hidden class (STRONG or not), should not matter... The point is that the _implementation class_  is a STRONG hidden class, so its lifetime gets tied to the caller's class loader.

**Review by JornVernee — APPROVED** [review]

**Pull request closed without merging**